Use asyncio instead of multiprocessing

aklajnert opened the pull request:
This PR aims to remove `multiprocessing` in favor of asyncio. It could still use some refactoring, but seem to work fine.  
I've left the `print`s for now, to show that the code is executed in parallel and the `poolSize` is respected.  

The change seems to not affect performance, and the code should be much easier to test and debug.

Asyncio code:  
```
real    2m52.298s
user    7m43.636s
sys     1m1.016s
```

Original code:  
```
real    2m52.413s
user    7m24.404s
sys     1m0.423s
```

Mikhail-Zex commented:
For how many songs at a time is this?

aklajnert commented:
@Mikhail-Zex - it uses the same `poolSize` parameter, so also 4.

Mikhail-Zex commented:
Agreed. How many songs in total? 4 in parallel but how many in total?

aklajnert commented:
I didn't pay attention, but I've used the album example from the help: https://open.spotify.com/album/2YMWspDGtbDgYULXvVQFM6?si=gF5dOQm8QUSo-NdZVsFjAQ
It has 27 songs. I've run `time python -m spotdl https://open.spotify.com/album/2YMWspDGtbDgYULXvVQFM6?si=gF5dOQm8QUSo-NdZVsFjAQ` command, removed all downloaded songs, and ran the same command again after switching branch to master. The time difference is basically none.

Mikhail-Zex commented:
Cool. I'll check it out and let you know.

Mikhail-Zex commented:
Hey @phcreery, do a review when free. It's practice for you too.

aklajnert commented:
The most important things are:
- the download process runs within a separate thread via `ThreadPoolExecutor`, as the library underneath is not asynchronous and would block the event loop
- ffmpeg is executed via `asyncio.subprocess` instead of `os.system()` to allow asynchronous operation.

I didn't test it for older Python versions, as I've used python 3.8. The code can still use some polishing.
The `download_song()` function has been moved into the `DownloadManager` class, as the `multiprocess` is not a problem anymore.

phcreery commented:
@aklajnert I like this! I tried messing around with asyncio but I guess I was doing it wrong! 

Your code runs great 👍

1. Since this is using a different library, is there a need for `ProgressRootProcess` (an extension of `multiprocessing.managers.BaseManager`)  or `AutoProxy`? It very well may be needed still but wanted to double check.


2. The second question, could the async function itself be asynchronous? For example, when the async function is called, the root script continues alongside the other "processes" to make the following work correctly?
```
self.loop.run_until_complete(asyncio.gather(*tasks))
print('Tasks dispatched and are running now')
```

3. @Mikhail-Zex likes to have full PEP8 docstings.

4. Althought the `download_song()` was originally intended to be used in the `DownloadManager` class, I think it was more useful as a function of its own. For example if a user wanted to use spotdl as a library but just wanted to download 1 song any given moment, the user would not need to utilize the `DownloadManager` class but could just use the `download_song()` function instead. Please let me know your opinions on this.

aklajnert commented:
I'll pay more attention to `1.` and `3.` when I'll be polishing the code.  
I'm not sure if I understand `2.` - the `await` keyword can be only run from an `async` function. The `loop.run_until_complete()` method spawns the tasks on the event loop and blocks until all tasks complete. If it wouldn't wait it would exit the program prematurely.

I think that instantiating `DownloadManager` is quite easy, and the `download_song()` function as a method lost 2 arguments that were required before and now are taken from the `DownloadManager` class.

Mikhail-Zex commented:
So @phcreery, what were you looking for during the review?

(@aklajnert I'm just using this as an example case to fine  tune reviewing I learnt over a few PR's with ritiek's guidance. Please don't mind)

Mikhail-Zex commented:
And we can actually use subprocess to spawn ffmpeg processes without using multiprocessing. And uvloop for nearly parallel input processing.

Uvloop is a faster version of Asyncio. 

See the second reply [here](https://stackoverflow.com/questions/27435284/multiprocessing-vs-multithreading-vs-asyncio-in-python-3) for more details

aklajnert commented:
Hmm, but the multiprocessing is already resolved on this branch, isn't it?  

I'm not sure if uvloop will give anything in that context. I did use uvloop on servers with hundreds of requests per second, and it indeed has great performance, but here, when we spawn 4 parallel tasks at a time we probably won't see any difference. And in general, I try to avoid third-party dependencies if they are not necessary. Especially modules that are written in C (as uvloop), as they tend to have problems on different OS. We can test if it gives any performance boost, but I doubt that.

In theory, we can limit download to 4 parallel tasks, and allow for unlimited `ffmpeg` processes which can give some performance boost. As far as I understand, the 4 tasks at a time constraint is to eliminate download errors, this doesn't apply to `ffmpeg`. However, I'm not sure if too many `ffmpeg` processes won't use too much PC resources that will make the process actually slower. Anyway - this approach will be a bit more tricky to implement.

Mikhail-Zex commented:
No not really. That's the most commonly reported error.

Ok. Let's stick with asyncio. 

ffmpeg is CPU intensive that's why we use multiprocessing - for speed and not really top eliminate errors. I run a 8core i7 7th gen and 4 simultaneous ffmpeg conversion drives up cpu usage to almost 90% across all cores.

aklajnert commented:
Makes sense, that's inline with my suspicions. Does the `ffmpeg` process utilize all cores, or is it one `ffmpeg` - one core? Maybe we can adjust the `poolSize` to the number of cores in a system? In general, I think it should be configurable.

aklajnert commented:
@phcreery @Mikhail-Zex - I've removed all remnants of `multuprocess`, and it still seems to work fine. I think the code is ready for final review.

Mikhail-Zex commented:
Cool. I'll do it by the weekend. 0lease switch from `master` to `next-rel-dev`

aklajnert commented:
@Mikhail-Zex - unfortunately, the merge conflicts are terrible. I'll work on it later and maybe will try to apply all the changes from this PR on the other branch from scratch.

Mikhail-Zex commented:
True but switch anyways, I'll do the integration once this PR has passed the final review.

aklajnert commented:
@Mikhail-Zex - I've fixed the conflicts and changed the base branch

Mikhail-Zex commented on spotdl/download/downloader.py:
```diff
@@ -325,15 +110,217 @@ def resume_download_from_tracking_file(self, trackingFilePath: str) -> None:
+            # ! downloadTrackers download queue and all is well...
+            # !
+            # ! None is again used as a convenient exit
+            remove(join(tempFolder, convertedFileName) + '.mp4')
```
split into 2 lines for readibility.

```
fileName = join(tempFolder, convertedFileName) + '.mp4'
remove(fileName)
```

It might seem like an insignificant gain but 100 insignificant gains makes an impact. similarly 100 insignificant let ups leads to un unmaintainable much (v2.2.2 all over again)

aklajnert replied:
that's not my code, but I'll do it

Mikhail-Zex reviewed (commented):
I might seem like a jerk at time in my comments but please forgive me for that, it's 2AM in the morning and I ain't getting sleep. That probably has something to do with it.

I have only gone through `downloader.py` and find only 2 problems.
1. Flouting of OOP Principles. You can read the refresher under spotDL/codeDocs.
2. Under utilization of comments. Very bad for maintainability.

Mikhail-Zex commented on spotdl/download/downloader.py:
```diff
@@ -251,20 +38,18 @@ class DownloadManager():
 
-        self.displayManager.clear()
+        # thread pool executor is used to run blocking code from a thread
+        self.thread_executor = concurrent.futures.ThreadPoolExecutor(max_workers=self.poolSize)
```
also, be clearer, I myself had to google `ThreadPoolExecutor` and read a bit about asyncIO (especially semaphores) to understand what you meant by "blocking code"

aklajnert replied:
> Import only what you need, not whole packages

I disagree. If you import the whole package, it is much more readable when used, as it doesn't force you to switch context and see where does it come from. See [Google style guide](https://google.github.io/styleguide/pyguide.html).  

As an example, when I was reading the code, I've noticed that the `ffmpeg` process was called via some mysterious `run_in_shell` function, which turned out to be renamed `os.system()`. If I'd see `os.system()` in place, I would immediately know what's going on there, instead of having to look for the `run_in_shell()` definition.

phcreery replied:
I agree with @aklajnert here

Mikhail-Zex replied:
Agreed. Lets continue with that then.

phcreery commented:
Other than @Mikhail-Zex comments, it looks good to me.

Mikhail-Zex commented:
Hey @phcreery, about your review from a few days back. It covers your doubts and clarifies some of the ideas behind the original implementation. Good work.

As to the few minor improvements, a review at least over here is meant to do 3 things: 
1. Keep it maintainable 
2. Definitely keep it maintainable 
3. Make sure we're giving the very best to our users

Taking about maintainability, you look for 2 things (after clearing your questions out, a reverse isn't necessarily a one shot one comment thing):
1. Simple architecture
2. Properly written readable code

Note that 'properly written' (OOPs and related ideas) and 'readable code' (good comments, and general readability)  are two separate points of interest. 

You can hassle one at a time but make it a point to comment your thoughts on both. And be explicit about your approval or disapproval. 

In this case, I won't be connecting on the architectural design, I trust your judgment on that.

aklajnert commented on spotdl/download/downloader.py:
```diff
@@ -251,20 +38,18 @@ class DownloadManager():
-        self.displayManager  = progressRoot.DisplayManager()
-        self.downloadTracker = progressRoot.DownloadTracker()
+        self.loop = asyncio.get_event_loop()
+        # semaphore is required to limit concurrent asyncio executions
```
Well, I've never seen such a commenting style, quite unusual, but I'll comply.

aklajnert commented on spotdl/download/downloader.py:
```diff
@@ -292,22 +82,17 @@ def download_multiple_songs(self, songObjList: List[SongObj]) -> None:
-            )
-        )
+        tasks = [self.pool_download(song) for song in songObjList]
+        self.loop.run_until_complete(asyncio.gather(*tasks))
```
Thanks, I was meant to extract it somewhere, but I've forgotten. Thanks for pointing that out.

aklajnert commented on spotdl/download/downloader.py:
```diff
@@ -325,15 +110,217 @@ def resume_download_from_tracking_file(self, trackingFilePath: str) -> None:
+
+        trackAudioStream = youtubeHandler.streams.get_audio_only()
+
+        task = functools.partial(self.perform_download, convertedFileName, tempFolder, trackAudioStream)
```
I've removed `functools`, it is not necessary here

aklajnert commented on spotdl/download/downloader.py:
```diff
@@ -325,15 +110,217 @@ def resume_download_from_tracking_file(self, trackingFilePath: str) -> None:
+        command = 'ffmpeg -v quiet -y -i "%s" -acodec libmp3lame -abr true -af loudnorm=I=-17 "%s"'
+        formattedCommand = command % (downloadedFilePath, convertedFilePath)
+
+        process = await asyncio.subprocess.create_subprocess_shell(formattedCommand)
```
How it would improve the readability?

aklajnert commented:
@Mikhail-Zex - I've fixed the code except for two points where I have doubts.  

I like your focus on maintainability, but I think you should also pay attention to other important parts of it like testing. How do you know that some bug you've fixed a year ago doesn't pop up again tomorrow? Do you test for everything manually?

Mikhail-Zex commented:
As of now yes. We had testing on our plan for release 3.2 but couldn't do so.

aklajnert commented:
In general it's much easier to test from start (maybe even with TDD) than adding tests to an existing code base. It shouldn't be delayed as it makes your code less reliable. Also other contributors can't be confident about changes in the code.

Mikhail-Zex commented:
Ok. What's TTD?

aklajnert commented:
Test-Driven Development. Some consider it a bit too extreme, but it's helpful to structure and design your code.
I recommend a book about it: https://www.obeythetestinggoat.com/, you can read it for free online.

Mikhail-Zex commented:
I read up then. Good resources go a long way in helping. THANKS.

phcreery commented:
Overall this looks much more straight-forward than the multiprocessing method. It is also very readable. 

The implementation of `run_in_executer` is quite clean but could use a little help. I recommend renaming `perform_download`  
 to something like `perform_audio_download` (to be more explicit) and moving its explanation to where it is called in `download_song`.

Go ahead and remove all the `print()` statements.

aklajnert commented:
> The implementation of `run_in_executer` is quite clean but could use a little help. I recommend renaming `perform_download`
> to something like `perform_audio_download` (to be more explicit) and moving its explanation to where it is called in `download_song`.

I've renamed the `_perform_download()` method, and moved the explanation into `_download_from_youtube()` instead `download_song()`. This nicely explains why the `run_in_executor()` has to be called, which isn't relevant for `download_song()` method as it is an implementation detail.

phcreery commented:
> I've renamed the `_perform_download()` method, and moved the explanation into `_download_from_youtube()` instead `download_song()`. This nicely explains why the `run_in_executor()` has to be called, which isn't relevant for `download_song()` method as it is an implementation detail.

Nicely done. @Mikhail-Zex what are your thoughts?

cristianpb commented:
I tried this branch to download a playlist and I got the following error: 

```
Fetching Playlist...
  6%|███████▉                                                                                                                                    |ETA: 11:25, 0.08min/songTraceback (most recent call last):
  File "/home/nano/.local/bin/spotdl", line 33, in <module>
    sys.exit(load_entry_point('spotdl', 'console_scripts', 'spotdl')())
  File "/home/nano/spotify-downloader/spotdl/__main__.py", line 102, in console_entry_point
    downloader.download_multiple_songs(songObjList)
  File "/home/nano/spotify-downloader/spotdl/download/downloader.py", line 83, in download_multiple_songs
    self._download_asynchronously(songObjList)
  File "/home/nano/spotify-downloader/spotdl/download/downloader.py", line 338, in _download_asynchronously
    self.loop.run_until_complete(asyncio.gather(*tasks))
  File "/usr/lib/python3.7/asyncio/base_events.py", line 587, in run_until_complete
    return future.result()
  File "/home/nano/spotify-downloader/spotdl/download/downloader.py", line 333, in _pool_download
    return await self.download_song(song_obj)
  File "/home/nano/spotify-downloader/spotdl/download/downloader.py", line 164, in download_song
    on_progress_callback=self.displayManager.pytube_progress_hook
  File "/home/nano/.local/lib/python3.7/site-packages/pytube/__main__.py", line 103, in __init__
    self.descramble()
  File "/home/nano/spotify-downloader/spotdl/patches/pyTube.py", line 149, in descramble
    self.player_config_args = get_ytplayer_config(self.watch_html)["args"]
  File "/home/nano/spotify-downloader/spotdl/patches/pyTube.py", line 241, in get_ytplayer_config
    raise RegexMatchError(
NameError: name 'RegexMatchError' is not defined
  6%|████████▋                                                                                                                                   |ETA: 14:57, 0.10min/songf
```

Maybe it's related with pytube ? 
I have `pytube (10.0.0)` and `pytube3 (9.6.4)`.

aklajnert commented:
@cristianpb - do you have the same error on `spotDL:next-rel-dev` branch?

Can you paste a full command (the playlist URL)?

Mikhail-Zex commented:
@cristianpb, force update pytube. pip can play up at times.

Mikhail-Zex commented:
Hey @aklajnert I was checking out your code today and 2 things. 

Extra long lines are bad for readability and as an extension for maintainability. Please limit line length to 100 characters.

I'm getting the RegxError too. I clean reinstalled python.

aklajnert commented:
> Extra long lines are bad for readability and as an extension for maintainability. Please limit line length to 100 characters.

Fixed, also added .editorconfig file that will help in some IDEs/Editors to visually identify too long lines by setting the right margin.

> I'm getting the RegxError too. I clean reinstalled python.

Now I see that I also have it. But the `next-rel-dev` branch has the same problem, so I don't think this is related to this PR. Also, it's quite weird, as I didn't update any dependency.

cristianpb commented:
I agree with @aklajnert , It's not related with this branch. 

I did a clean install and tested `spotDL:next-rel-dev` and `aklajnert:switch_from_multiprocess_to_asyncio` with similar RegexMatchError pytube problems.

aklajnert commented:
@MikhailZex - I've changed the base to `master`, but I didn't have time to test it yet. Given the fact that the merge was quite nasty, it'll be good to test it thoroughly. I've created another branch `aklajnert:switch_to_asyncio_next_rel_dev` with the changes before switching base in case it will be needed in the future.
Additionally, this branch has still an unresolved discussion.

MikhailZex commented:
So should I merge it or wait a while? (Can't review/ test code, no lap)

MikhailZex commented:
Also, I sent you an org invite, join in anyways.

aklajnert commented:
> So should I merge it or wait a while? (Can't review/ test code, no lap)

Definitely wait until it will be tested - I can't do that right now. Also, an additional look at the code could be helpful to see if I didn't mess up anything while merging.

MikhailZex commented:
@spotDL/joiners try testing this.

aklajnert commented:
OK, I've messed up and it doesn't work. I'll reapply the changes again on master in a bit different way.

aklajnert commented:
OK, now it should be fine. I've identified a problem with Python 3.7 on Windows but managed to fix it.
It seems to work fine for songs, albums, and playlists on Windows and Linux (WSL).

I had to switch the `pytube` version, which basically made me reimplement #1015 here to make it work.

Sometimes `youtubeHandler.streams.get_audio_only()` function returns `None` and the song is not downloaded. I don't know why and I don't have more time to investigate. Running the program again, with the same arguments fixes the problem. Does that ring any bell?

Would be good if someone else could also test it out.

phcreery commented:
I can try to test it out later this evening. I will try to let you know how it goes.

phcreery commented:
I just tested this again it runs without problems. I am not getting the RegexMatchError pytube error.

aklajnert commented:
I think merging this branch will cause no harm - it works definitely better than the current master, as it doesn't raise an exception.

MikhailZex commented:
Cool. Could you upload the package to pypi too? If you have an account. I'll send you access.
